crates/gpui/src/platform/windows/direct_write.rs

```diff
@@ -0,0 +1,1286 @@
+        };
+        let brush_property = D2D1_BRUSH_PROPERTIES {
+            opacity: 1.0,
+            transform: Matrix3x2 {
+                M11: params.scale_factor,
+                M12: 0.0,
+                M21: 0.0,
+                M22: params.scale_factor,
+                M31: 0.0,
+                M32: 0.0,
+            },
+        };
+
+        let total_bytes;
+        let bitmap_format;
+        let render_target_property;
+        let bitmap_stride;
+        if params.is_emoji {
+            total_bytes = bitmap_size.height.0 as usize * bitmap_size.width.0 as usize * 4;
+            bitmap_format = &GUID_WICPixelFormat32bppPRGBA;
+            render_target_property = D2D1_RENDER_TARGET_PROPERTIES {
+                r#type: D2D1_RENDER_TARGET_TYPE_DEFAULT,
+                pixelFormat: D2D1_PIXEL_FORMAT {
+                    format: DXGI_FORMAT_R8G8B8A8_UNORM,
+                    alphaMode: D2D1_ALPHA_MODE_PREMULTIPLIED,
+                },
+                dpiX: params.scale_factor * 96.0,
+                dpiY: params.scale_factor * 96.0,
+                usage: D2D1_RENDER_TARGET_USAGE_NONE,
+                minLevel: D2D1_FEATURE_LEVEL_DEFAULT,
+            };
+            bitmap_stride = bitmap_size.width.0 as u32 * 4;
+        } else {
+            total_bytes = bitmap_size.height.0 as usize * bitmap_size.width.0 as usize;
+            bitmap_format = &GUID_WICPixelFormat8bppAlpha;
+            render_target_property = D2D1_RENDER_TARGET_PROPERTIES {
+                r#type: D2D1_RENDER_TARGET_TYPE_DEFAULT,
+                pixelFormat: D2D1_PIXEL_FORMAT {
+                    format: DXGI_FORMAT_A8_UNORM,
+                    alphaMode: D2D1_ALPHA_MODE_STRAIGHT,
+                },
+                dpiX: params.scale_factor * 96.0,
+                dpiY: params.scale_factor * 96.0,
+                usage: D2D1_RENDER_TARGET_USAGE_NONE,
+                minLevel: D2D1_FEATURE_LEVEL_DEFAULT,
+            };
+            bitmap_stride = bitmap_size.width.0 as u32;
+        }
+
+        unsafe {
+            let bitmap = self.components.bitmap_factory.CreateBitmap(
+                bitmap_size.width.0 as u32,
+                bitmap_size.height.0 as u32,
+                bitmap_format,
+                WICBitmapCacheOnLoad,
+            )?;
+            let render_target = self
+                .components
+                .d2d1_factory
+                .CreateWicBitmapRenderTarget(&bitmap, &render_target_property)?;
+            let brush = render_target.CreateSolidColorBrush(&BRUSH_COLOR, Some(&brush_property))?;
+            let subpixel_shift = params
+                .subpixel_variant
+                .map(|v| v as f32 / SUBPIXEL_VARIANTS as f32);
+            let baseline_origin = D2D_POINT_2F {
+                x: subpixel_shift.x / params.scale_factor,
+                y: subpixel_shift.y / params.scale_factor,
+            };
+
+            let render_target: ID2D1DeviceContext4 = std::mem::transmute(render_target);
```

Comment:
As well as for here, though if this is a valid operation let's include a comment explaining it.

Reply:
These two structs actually are the same (Microsoft said this), and Microsoft enforces type conversion on these two structs in their project.

At https://github.com/microsoft/terminal/blob/main/src/renderer/atlas/BackendD2D.cpp#L86 Microsfot said: "ID2D1RenderTarget and ID2D1DeviceContext are the same..."